File: DSA/Java/src/lib/lists/arrays/Array.java
package lib.lists.arrays;

import lib.lists.linkedlists.SinglyLinkedList;
import nodes.BinaryTreeNode;
import nodes.ListNode;

public class Array {
    private final int[] array;
    public final int length;

    public Array(int[] array) {
        this.array = array;
        this.length = this.array.length;
    }

    public ListNode toSinglyLinkedList() {
        ListNode dummyNode = new ListNode(0);
        ListNode head = new ListNode(this.array[0]);
        dummyNode.next = head;

        for (int i = 1; i < this.length; i++) {
            head.next = new ListNode(this.array[i]);
            head = head.next;
        }

        return dummyNode.next;
    }

    /**
     * To represent a cycle in the given linked list, we use an integer pos which represents
     * the position (0-indexed) in the linked list where tail connects to. If pos is -1, then
     * there is no cycle in the linked list.
     */
    public ListNode toCyclicSinglyLinkedList(int pos) {
        ListNode head = this.toSinglyLinkedList();

        if (pos != -1) {
            SinglyLinkedList ls = new SinglyLinkedList(head);
            ListNode tail = ls.getNodeByIndex(ls.size() - 1);
            tail.next = ls.getNodeByIndex(pos);
        }

        return head;
    }

    public BinaryTreeNode toBinaryTree() {
        BinaryTreeNode root = new BinaryTreeNode(this.array[0]);
        generateTreeFromArray(this.array, this.length, root, 0);
        return root;
    }

    private static void generateTreeFromArray(int[] array, int n, BinaryTreeNode root, int i) {
        int lpos = 2 * i + 1, rpos = 2 * i + 2;
        if (lpos < n && array[lpos] != 0) {
            root.left = new BinaryTreeNode(array[lpos]);
            generateTreeFromArray(array, n, root.left, lpos);
        }
        if (rpos < n && array[rpos] != 0) {
            root.right = new BinaryTreeNode(array[rpos]);
            generateTreeFromArray(array, n, root.right, rpos);
        }
    }
}

File: DSA/Cpp/src/dsa/lib/arrays/array.cpp
#include "array.h"

namespace dsa::lib::arrays {

Array::Array(int *ary, length_t length) : ary(ary), length(length) {}

ListNode *Array::toSinglyLinkedList() const {
    return this->toCyclicSinglyLinkedList(-1);
}

ListNode *Array::toCyclicSinglyLinkedList(int pos) const {
    if (!this->length || !*this->ary) return nullptr;
    ListNode dummy;
    auto *head = new ListNode(*this->ary);
    dummy.next = head;
    ListNode *cycleEntryNode = head;

    for (int i = 1; i < this->length; i++) {
        auto *node = new ListNode(*(this->ary + i));
        head->next = node;
        if (i == pos) cycleEntryNode = node;
        head = head->next;
    }

    if (pos >= 0) head->next = cycleEntryNode;

    return dummy.next;
}

BinaryTreeNode *Array::toBinaryTree() const {
    if (!this->length || !*this->ary) return nullptr;
    auto *root = new BinaryTreeNode(*this->ary);
    generateTreeFromArray(this->ary, this->length, root, 0);
    return root;
}
void Array::generateTreeFromArray(int *ary, length_t length, BinaryTreeNode *root, pos_t i) {
    pos_t lpos = 2 * i + 1, rpos = 2 * i + 2;
    if (lpos < length && *(ary + lpos)) {
        root->left = new BinaryTreeNode(*(ary + lpos));
        generateTreeFromArray(ary, length, root->left, lpos);
    }
    if (rpos < length && *(ary + rpos)) {
        root->right = new BinaryTreeNode(*(ary + rpos));
        generateTreeFromArray(ary, length, root->right, rpos);
    }
}

}  // namespace dsa::lib::arrays